capsules/src/mac.rs

```diff
@@ -0,0 +1,365 @@
+//! Implements 802.15.4 MAC device functionality as an abstraction layer over
+//! the raw radio transceiver hardware. The abstraction difference between a MAC
+//! device and a raw radio transceiver is that the MAC devices exposes a
+//! frame-oriented interface to its users, whereas the radio transceiver
+//! transmits raw byte sequences. There is some abstraction breaking here,
+//! though because the following are still implemented at the hardware level:
+//! - CSMA-CA backoff
+//! - FCS generation and verification
+//!
+//! TODO: Encryption/decryption
+//! TODO: Sending beacon frames
+//! TODO: Channel scanning
+
+use core::cell::Cell;
+use kernel::ReturnCode;
+use kernel::common::take_cell::TakeCell;
+use kernel::hil::radio;
+use net::ieee802154::*;
+
+#[derive(Copy, Clone, Eq, PartialEq, Debug)]
+pub struct FrameInfo {
+    // These offsets are relative to buf[radio::PSDU_OFFSET..] so that
+    // the mac frame length is data_offset + data_len
+    mac_payload_offset: usize,
+    data_offset: usize,
+    data_len: usize,
+    security_params: Option<(Security, [u8; 16])>,
+}
+
+impl FrameInfo {
+    pub fn frame_length(&self) -> usize {
+        self.data_offset + self.data_len
+    }
+
+    pub fn remaining_data_capacity(&self, buf: &[u8]) -> usize {
+        buf.len() - radio::PSDU_OFFSET - radio::MFR_SIZE - self.data_offset - self.data_len
+    }
+
+    pub fn append_payload(&mut self, buf: &mut [u8], payload: &[u8]) -> ReturnCode {
+        if payload.len() > self.remaining_data_capacity(buf.as_ref()) {
+            return ReturnCode::ENOMEM;
+        }
+        let begin = radio::PSDU_OFFSET + self.data_offset + self.data_len;
+        buf[begin..begin + payload.len()].copy_from_slice(payload);
+        self.data_len += payload.len();
+
+        ReturnCode::SUCCESS
+    }
+}
+
+pub trait Mac {
```

Comment:
I've documented the `get_*` and `set_*` methods. The idea is that after you call one or more `set_*` functions, you must always call `config_commit` once, otherwise there is no guarantee that the RF233 registers match the MAC software layer's addressing information. I've also made it so that calling `config_commit` when a configuration process is still ongoing will effectively "schedule" another config. This is because the underlying RF233 driver treats a configuration as an atomic operation that cannot be interrupted/restarted, so if new addressing parameters are provided, the configuration process must be repeated.